android-sdk/src/main/java/ai/evolv/android_sdk/EvolvContextImpl.java

```diff
@@ -168,4 +189,242 @@ private void ensureInitialized() {
+        ensureInitialized();
+        return remoteContext.has(key) || localContext.has(key);
+    }
+
+    // TODO: 26.07.2021 need to test!
+    @Override
+    public void update(JsonObject update, boolean local) {
+        ensureInitialized();
+        JsonObject context = local ? localContext : remoteContext;
+        flatten(update);
+
+        JsonObject flattenedBefore = new JsonObject();
+
+        for (String key : flattened.keySet()) {
+            flattenedBefore.add(key,context.get(key));
+        }
+
+        if (local) {
+
+            for (Map.Entry<String, JsonElement> entry : update.entrySet()) {
+                localContext.add(entry.getKey(),entry.getValue());
+            }
+            context = localContext;
+        } else {
+            for (Map.Entry<String, JsonElement> entry : update.entrySet()) {
+                remoteContext.add(entry.getKey(),entry.getValue());
+            }
+            context = remoteContext;
+        }
+
+        JsonObject updated = this.resolve();
+
+        for (String key : flattened.keySet()) {
+
+            if (flattenedBefore.get(key) == null) {
+
+                JsonObject object = new JsonObject();
+                object.addProperty("type",CONTEXT_INITIALIZED);
+                object.addProperty("key",key);
+                object.add("flattened_key",flattened.get(key));
+                object.addProperty("local",local);
+                object.add("updated",updated);
+
+                waitForIt.emit(this, CONTEXT_VALUE_ADDED, object);
+            }else if (flattenedBefore.get(key) != context.get(key)) {
+
+                JsonObject object = new JsonObject();
+                object.addProperty("type",CONTEXT_INITIALIZED);
+                object.addProperty("key",key);
+                object.add("flattened_key",flattened.get(key));
+                object.add("flattenedBefore_key",flattenedBefore.get(key));
+                object.addProperty("local",local);
+                object.add("updated",updated);
+
+                waitForIt.emit(this, CONTEXT_VALUE_CHANGED, object);
+            }
+        }
+    }
+
+    private void flatten(JsonElement map) {
+        recurseflatten(map, "");
+    }
+
+    private String recurseflatten(JsonElement current, String parentKey) {
+        Set<String> keys = current.getAsJsonObject().keySet();
+
+        Iterator<String> iterator = keys.iterator();
+        Set<String> items = new HashSet<>();
+
+        while (iterator.hasNext()) {
+            String key = iterator.next();
+
+            if (startsWithFilter.apply(key)) {
+                JsonElement element = current.getAsJsonObject().get(key);
+                if (element.isJsonObject()) {
+                    if (element.getAsJsonObject().size() == 0) continue;
+                }
+                String newKey = !parentKey.isEmpty() ? (parentKey + '.' + key) : key;
+                items.add(newKey);
+                flattened.addProperty(newKey,newKey);
```

Comment:
This side effect looks a bit risky for holding onto the state - would this be better as the return value?

Reply:
this is not a problem, because the method, when calling itself, will add the value to the object (which is what we need). But if necessary, I will try to rewrite with a return value